aten/src/ATen/native/cuda/DilatedMaxPool2d.cu

```diff
@@ -229,34 +248,42 @@ __global__ void max_pool_backward_nhwc(const int nthreads, const scalar_t* top_d
   int iendW = ::min(istartW+iW, width);
 
   for (int ih = istartH; ih < iendH; ih+=blockDim.z) {
+    int phstart = p_start(ih, pad_h, kernel_h, dilation_h, stride_h);
+    int phend = p_end(ih, pad_h, pooled_height, stride_h);
     for (int iw = istartW; iw < iendW; iw+=blockDim.y) {
-      int phstart = p_start(ih, pad_h, kernel_h, dilation_h, stride_h);
-      int phend = p_end(ih, pad_h, pooled_height, stride_h);
       int pwstart = p_start(iw, pad_w, kernel_w, dilation_w, stride_w);
       int pwend = p_end(iw, pad_w, pooled_width, stride_w);
       if ((phstart + 1 != phend) || (pwstart + 1 != pwend)) {
         for(int oh = phstart; oh < phend; ++oh) {
           for(int ow = pwstart; ow < pwend; ++ow) {
+            int cached_index = threadIdx.x; 
             const int64_t* ptr_top_mask = top_mask + oh * out_stride_h + ow * out_stride_w;
-            for (int c = threadIdx.x; c < channels; c += blockDim.x) {
-              if (ptr_top_mask[c] == ih * width + iw) {
-                out_cached[c] += scalar_cast<scalar_t>(top_diff[oh * out_stride_h + ow * out_stride_w + c]);
+            for (int c = channel_offset; c < channels; c += blockDim.x*kernel_stride_C) {
+              if (ptr_top_mask[c*out_stride_c] == ih * width + iw) {
```

Comment:
NITPICK: `ih * width + iw` is reused in the for_loop(c). We can probably reuse it if you are not under high register pressure.

Reply:
Thanks, added.